File: 01-basic-auth/node/src/delivery/authHandler.ts
import { Request, Response } from 'express';
import { AuthUsecase } from '../usecase/authUsecase';

/**
 * HTTP handlers (Delivery слой) для Express.
 *
 * Контракт:
 *   POST   /api/v1/auth/register  body: { email, password } → 201 { message, user }
 *   POST   /api/v1/auth/login     body: { email, password } → 200 { message, user }
 *   DELETE /api/v1/auth/delete    body: { email, password } → 200 { message }
 *
 * Handler'ы читают тело запроса, вызывают use case и формируют HTTP-ответ.
 * Basic Auth middleware больше не используется.
 */
export function createAuthHandler(authUsecase: AuthUsecase) {
  return {
    // POST /api/v1/auth/register
    // Создаёт нового пользователя. Возвращает 201 и данные пользователя.
    async register(req: Request, res: Response) {
      const { email, password } = req.body;
      if (!email || !password) {
        res.status(400).json({ error: 'email and password required' });
        return;
      }
      try {
        const user = await authUsecase.register(email, password);
        res.status(201).json({
          message: 'Пользователь успешно зарегистрирован',
          user,
        });
      } catch (err: any) {
        if (err.message === 'user already exists') {
          res.status(409).json({ error: err.message });
        } else {
          res.status(500).json({ error: err.message });
        }
      }
    },

    // POST /api/v1/auth/login
    // Проверяет email + пароль. Возвращает 200 и данные пользователя.
    async login(req: Request, res: Response) {
      const { email, password } = req.body;
      if (!email || !password) {
        res.status(400).json({ error: 'email and password required' });
        return;
      }
      try {
        const user = await authUsecase.login(email, password);
        const publicUser = {
          id: user.id,
          email: user.email,
          created_at: user.createdAt.toISOString(),
        };
        res.status(200).json({
          message: 'Добро пожаловать!',
          user: publicUser,
        });
      } catch (err: any) {
        if (err.message === 'invalid email or password') {
          res.status(401).json({ error: err.message });
        } else {
          res.status(500).json({ error: err.message });
        }
      }
    },

    // DELETE /api/v1/auth/delete
    // Принимает email + пароль в теле — проверяет и удаляет аккаунт.
    async deleteByCredentials(req: Request, res: Response) {
      const { email, password } = req.body;
      if (!email || !password) {
        res.status(400).json({ error: 'email and password required' });
        return;
      }
      try {
        await authUsecase.deleteByCredentials(email, password);
        res.status(200).json({ message: 'Пользователь успешно удалён' });
      } catch (err: any) {
        if (err.message === 'invalid email or password') {
          res.status(401).json({ error: err.message });
        } else {
          res.status(500).json({ error: err.message });
        }
      }
    },
  };
}

File: 01-basic-auth/go/internal/delivery/auth_handler.go
package delivery

// HTTP handlers для авторизации (Delivery слой).
//
// Handler'ы — это "входная точка" HTTP:
// - читают и валидируют входные данные (JSON/body)
// - вызывают бизнес-логику (UseCase)
// - формируют HTTP-ответ (status code + JSON)
import (
	"basic-auth/internal/delivery/middleware"
	"basic-auth/internal/usecase"
	"encoding/json"
	"net/http"
)

type AuthHandler struct {
	authUsecase *usecase.AuthUsecase
}

// NewAuthHandler связывает HTTP-слой с UseCase-слоем.
func NewAuthHandler(authUsecase *usecase.AuthUsecase) *AuthHandler {
	return &AuthHandler{authUsecase: authUsecase}
}

type RegisterRequest struct {
	Email    string `json:"email"`
	Password string `json:"password"`
}

// RegisterHandler — публичный endpoint регистрации.
// При успехе возвращает 201 Created и пользователя (без password).
func (h *AuthHandler) RegisterHandler(w http.ResponseWriter, r *http.Request) {
	var req RegisterRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	user, err := h.authUsecase.Register(req.Email, req.Password)
	if err != nil {
		if err.Error() == "user already exists" {
			http.Error(w, err.Error(), http.StatusConflict)
			return
		}
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(user)
}

// MeHandler — защищённый endpoint "кто я".
// user_id берётся из context: туда его положил Basic Auth middleware после успешной проверки.
func (h *AuthHandler) MeHandler(w http.ResponseWriter, r *http.Request) {
	userID, ok := r.Context().Value(middleware.UserIDKey).(string)
	if !ok || userID == "" {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}

	user, err := h.authUsecase.GetUserByID(userID)
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(user)
}

// DeleteUserHandler — защищённый endpoint удаления своего аккаунта.
func (h *AuthHandler) DeleteUserHandler(w http.ResponseWriter, r *http.Request) {
	userID, ok := r.Context().Value(middleware.UserIDKey).(string)
	if !ok || userID == "" {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}

	err := h.authUsecase.DeleteUserById(userID)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}
